### PR TITLE
Add openapi generation for the kube-aggregator

### PR DESCRIPTION
**Pull request by bjrara**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>

/kind api-change

> /kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

Today the kube-aggregator OpenAPI definitions are embedded in the master apiserver. 

We want to seperate the OpenAPI generation for the kube-aggregator to a new package so that we can use this pure definition list in the kube aggregator in the standalone mode.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by lavalamp** [comment]
/assign @deads2k 

David may have a stronger opinion than I do about this

**Comment by lavalamp** [comment]
I think this is probably fine... we may be missing a test of some sort to make sure this binary keeps building / working.

**Comment by lavalamp** [comment]
It looks like you need to sign a CLA before we can accept this...

**Comment by bjrara** [comment]
> I think this is probably fine... we may be missing a test of some sort to make sure this binary keeps building / working.

The check is added `pull-kubernetes-files-remake` test. It ensures the openapi file will be generated.

**Comment by lavalamp** [comment]
/approve
/lgtm

(you'll need to edit the description to add a release note before it can merge-- I think "NONE" is fine)

**Comment by bjrara** [comment]
> /approve
> /lgtm
> 
> (you'll need to edit the description to add a release note before it can merge-- I think "NONE" is fine)

Updated. Thanks for the reminder.

**Comment by liggitt** [comment]
I expected the new generated file to be checked in ... we check in `apiextensions-apiserver/pkg/generated/openapi/zz_generated.openapi.go` because we're missing the generator scripts and want the published module to be complete

File: api/api-rules/aggregator_violation_exceptions.list [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+API rule violation: list_type_missing,k8s.io/apimachinery/pkg/apis/meta/v1,RootPaths,Paths
+API rule violation: list_type_missing,k8s.io/apimachinery/pkg/apis/meta/v1,StatusDetails,Causes
+API rule violation: list_type_missing,k8s.io/apimachinery/pkg/apis/meta/v1,Table,ColumnDefinitions
+API rule violation: list_type_missing,k8s.io/apimachinery/pkg/apis/meta/v1,Table,Rows
+API rule violation: list_type_missing,k8s.io/apimachinery/pkg/apis/meta/v1,TableRow,Cells
+API rule violation: list_type_missing,k8s.io/apimachinery/pkg/apis/meta/v1,TableRow,Conditions
+API rule violation: list_type_missing,k8s.io/apimachinery/pkg/apis/meta/v1,UpdateOptions,DryRun
+API rule violation: list_type_missing,k8s.io/apimachinery/pkg/runtime,RawExtension,Raw
+API rule violation: list_type_missing,k8s.io/apimachinery/pkg/runtime,Unknown,Raw
+API rule violation: list_type_missing,k8s.io/kube-aggregator/pkg/apis/apiregistration/v1,APIServiceSpec,CABundle
```
Comment (liggitt) [inline api/api-rules/aggregator_violation_exceptions.list:27]: should be atomic?

File: api/api-rules/aggregator_violation_exceptions.list [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+API rule violation: list_type_missing,k8s.io/kube-aggregator/pkg/apis/apiregistration/v1,APIServiceStatus,Conditions
```
Comment (liggitt) [inline api/api-rules/aggregator_violation_exceptions.list:28]: should be listType=map, with type as the map key?

File: api/api-rules/aggregator_violation_exceptions.list [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+API rule violation: list_type_missing,k8s.io/kube-aggregator/pkg/apis/apiregistration/v1beta1,APIServiceSpec,CABundle
```
Comment (liggitt) [inline api/api-rules/aggregator_violation_exceptions.list:29]: atomic?

File: api/api-rules/aggregator_violation_exceptions.list [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+API rule violation: list_type_missing,k8s.io/kube-aggregator/pkg/apis/apiregistration/v1beta1,APIServiceStatus,Conditions
```
Comment (liggitt) [inline api/api-rules/aggregator_violation_exceptions.list:30]: should be listType=map, with type as the map key?

**Review by liggitt — COMMENTED** [review]
is this also related to https://github.com/kubernetes/kubernetes/issues/89264?

**Comment by bjrara** [comment]
> I expected the new generated file to be checked in and used ... we check in `apiextensions-apiserver/pkg/generated/openapi/zz_generated.openapi.go` because we're missing the generator scripts and want the published module to be complete

This PR is a prerequisite of PR https://github.com/kubernetes/kubernetes/pull/91525. The OPENAPI configs used by the standalone kube-aggregator should contain and only contain OPENAPI configs required by the kube-aggregator pkg.

File: api/api-rules/aggregator_violation_exceptions.list (hunk 4 above)
Comment (bjrara) [reply api/api-rules/aggregator_violation_exceptions.list:30]: > is this also related to #89264?

I believe this PR, together with the PR https://github.com/kubernetes/kubernetes/pull/91525  will solve this issue.

**Comment by bjrara** [comment]
> is this also related to #89264?

I believe this PR, together with https://github.com/kubernetes/kubernetes/pull/91525  will solve this issue.

File: api/api-rules/aggregator_violation_exceptions.list (hunk 1 above)
Comment (bjrara) [reply api/api-rules/aggregator_violation_exceptions.list:27]: I don't think this field is marked as atomic in the openapi. Would you mind another check?

File: api/api-rules/aggregator_violation_exceptions.list (hunk 2 above)
Comment (bjrara) [reply api/api-rules/aggregator_violation_exceptions.list:28]: The APIServiceStatus doesn't have the new api annotations. It is using `patchStrategy` and `patchMergeKey` markers for the compatibility.

File: api/api-rules/aggregator_violation_exceptions.list (hunk 3 above)
Comment (bjrara) [reply api/api-rules/aggregator_violation_exceptions.list:29]: ^^^

File: api/api-rules/aggregator_violation_exceptions.list (hunk 4 above)
Comment (bjrara) [reply api/api-rules/aggregator_violation_exceptions.list:30]: ^^^

File: api/api-rules/aggregator_violation_exceptions.list (hunk 1 above)
Comment (yue9944882) [reply api/api-rules/aggregator_violation_exceptions.list:27]: > is this also related to #89264?

the default list-type is atmoic if it's not annotated, is it? literally every array fields from built-in resources has the `list_type_missing` violation (but under different files). i suppose these violation reports are trival

**Review by yue9944882 — COMMENTED** [review]
another odd thing i found after applying the patch and re-gen the openapi locally is that there're duplicated entries between `pkg/generated/openapi/zz_generated.openapi.go` and the new `staging/src/k8s.io/kube-aggregator/pkg/generated/openapi/zz_generated.openapi.go` involved by this pull (we didn't have duplication from crd-apiserver's generated openapi before):

```
		"k8s.io/kube-aggregator/pkg/apis/apiregistration/v1.APIService":                                               schema_pkg_apis_apiregistration_v1_APIService(ref),
		"k8s.io/kube-aggregator/pkg/apis/apiregistration/v1.APIServiceCondition":                                      schema_pkg_apis_apiregistration_v1_APIServiceCondition(ref),
		"k8s.io/kube-aggregator/pkg/apis/apiregistration/v1.APIServiceList":                                           schema_pkg_apis_apiregistration_v1_APIServiceList(ref),
		"k8s.io/kube-aggregator/pkg/apis/apiregistration/v1.APIServiceSpec":                                           schema_pkg_apis_apiregistration_v1_APIServiceSpec(ref),
		"k8s.io/kube-aggregator/pkg/apis/apiregistration/v1.APIServiceStatus":     
```
however, the openapi aggregator seems to dedup these entries if we download from /openapi endpoint.

/cc @roycaihw

**Comment by bjrara** [comment]
> another odd thing i found after applying the patch and re-gen the openapi locally is that there're duplicated entries between `pkg/generated/openapi/zz_generated.openapi.go` and the new `staging/src/k8s.io/kube-aggregator/pkg/generated/openapi/zz_generated.openapi.go` involved by this pull (we didn't have duplication from crd-apiserver's generated openapi before):
> 
> ```
> 		"k8s.io/kube-aggregator/pkg/apis/apiregistration/v1.APIService":                                               schema_pkg_apis_apiregistration_v1_APIService(ref),
> 		"k8s.io/kube-aggregator/pkg/apis/apiregistration/v1.APIServiceCondition":                                      schema_pkg_apis_apiregistration_v1_APIServiceCondition(ref),
> 		"k8s.io/kube-aggregator/pkg/apis/apiregistration/v1.APIServiceList":                                           schema_pkg_apis_apiregistration_v1_APIServiceList(ref),
> 		"k8s.io/kube-aggregator/pkg/apis/apiregistration/v1.APIServiceSpec":                                           schema_pkg_apis_apiregistration_v1_APIServiceSpec(ref),
> 		"k8s.io/kube-aggregator/pkg/apis/apiregistration/v1.APIServiceStatus":     
> ```
> 
> however, the openapi aggregator seems to dedup these entries if we download from /openapi endpoint.
> 
> /cc @roycaihw

IIUC, the `staging/src/k8s.io/kube-aggregator/**/zz_generated.openapi.go` is for the openapi config setup of the standalone kube-aggregator.

The `staging/src/k8s.io/apiextensions-apiserver/**/zz_generated.openapi.go` is for the openapi config setup of the standalone apiextensions.

That's why APIService will not be generated in the apiextensions openapi, because the standalone apiextentions server doesn't have these types.

The kube-apiserver uses its own openapi config file which is `pkg/generated/openapi/zz_generated.openapi.go`. It contains a complete list of all built-in types. 

The usage can be found as follow:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-apiserver/app/server.go#L457

These three generated are independent in terms of usage.

**Comment by lavalamp** [comment]
I think you need to regenerate something, but just in case:

/test pull-kubernetes-verify

File: api/api-rules/aggregator_violation_exceptions.list (hunk 2 above)
Comment (liggitt) [reply api/api-rules/aggregator_violation_exceptions.list:28]: please add the listType annotations to resolve these warnings for pkg/apis/apiregistration items

Comment (bjrara) [reply api/api-rules/aggregator_violation_exceptions.list]: done

File: api/api-rules/aggregator_violation_exceptions.list (hunk 1 above)
Comment (bjrara) [reply api/api-rules/aggregator_violation_exceptions.list]: done

File: staging/src/k8s.io/kube-aggregator/pkg/apis/apiregistration/types.go [hunk 5]
```diff
@@ -60,6 +60,8 @@ type APIServiceSpec struct {
 	InsecureSkipTLSVerify bool
 	// CABundle is a PEM encoded CA bundle which will be used to validate an API server's serving certificate.
 	// If unspecified, system trust roots on the apiserver are used.
+	// +listType=map
+	// +listMapKey=type
```
Comment (liggitt) [inline staging/src/k8s.io/kube-aggregator/pkg/apis/apiregistration/types.go:64]:
```suggestion
	// +listType=atomic
```

File: staging/src/k8s.io/kube-aggregator/pkg/apis/apiregistration/v1/types.go [hunk 6]
```diff
@@ -60,6 +60,8 @@ type APIServiceSpec struct {
 	InsecureSkipTLSVerify bool `json:"insecureSkipTLSVerify,omitempty" protobuf:"varint,4,opt,name=insecureSkipTLSVerify"`
 	// CABundle is a PEM encoded CA bundle which will be used to validate an API server's serving certificate.
 	// If unspecified, system trust roots on the apiserver are used.
+	// +listType=map
+	// +listMapKey=type
```
Comment (liggitt) [inline staging/src/k8s.io/kube-aggregator/pkg/apis/apiregistration/v1/types.go:64]:
```suggestion
	// +listType=atomic
```

File: staging/src/k8s.io/kube-aggregator/pkg/apis/apiregistration/v1beta1/types.go [hunk 7]
```diff
@@ -63,6 +63,8 @@ type APIServiceSpec struct {
 	InsecureSkipTLSVerify bool `json:"insecureSkipTLSVerify,omitempty" protobuf:"varint,4,opt,name=insecureSkipTLSVerify"`
 	// CABundle is a PEM encoded CA bundle which will be used to validate an API server's serving certificate.
 	// If unspecified, system trust roots on the apiserver are used.
+	// +listType=map
+	// +listMapKey=type
```
Comment (liggitt) [inline staging/src/k8s.io/kube-aggregator/pkg/apis/apiregistration/v1beta1/types.go:67]:
```suggestion
	// +listType=atomic
```

**Comment by bjrara** [comment]
@liggitt Thanks for your suggestions on the openapi spec. Files regenerated.

**Pull request merged**